apps/prairielearn/src/lib/assessment.js

```diff
@@ -40,482 +42,417 @@ const InstanceLogSchema = z.object({
-            { grading_job_id: content.gradingId },
-            (err, result) => {
-              if (ERR(err, callback)) return;
-              let assessment_instance_id = result.rows[0].assessment_instance_id;
-              ltiOutcomes.updateScore(assessment_instance_id, (err) => {
-                if (ERR(err, callback)) return;
-                callback(null);
-              });
-            },
-          );
-        },
-      ]);
-    } finally {
-      externalGradingSocket.gradingJobStatusUpdated(content.gradingId);
+/**
+ * Process the result of an external grading job.
+ *
+ * @param {Object} content - The grading job data to process.
+ */
+export async function processGradingResult(content) {
+  try {
+    if (!_.isObject(content.grading)) {
+      throw error.makeWithData('invalid grading', { content: content });
     }
-  },
-
-  /**
-   * Updates statistics for all assessments in a course instance, but only if an
-   * update is needed.
-   *
-   * @param {number} course_instance_id - The course instance ID.
-   */
-  async updateAssessmentStatisticsForCourseInstance(course_instance_id) {
-    const result = await sqldb.queryAsync(sql.select_assessments_for_statistics_update, {
-      course_instance_id,
-    });
-    await async.eachLimit(result.rows, 3, async (row) => {
-      await module.exports.updateAssessmentStatistics(row.assessment_id);
-    });
-  },
-
-  /**
-   * Updates statistics for an assessment, if needed.
-   *
-   * @param {number} assessment_id - The assessment ID.
-   */
-  async updateAssessmentStatistics(assessment_id) {
-    await sqldb.runInTransactionAsync(async () => {
-      // lock the assessment
-      await sqldb.queryOneRowAsync(sql.select_assessment_lock, { assessment_id });
-
-      // check whether we need to update the statistics
-      const result = await sqldb.queryOneRowAsync(sql.select_assessment_needs_statisics_update, {
-        assessment_id,
-      });
-      if (!result.rows[0].needs_statistics_update) return;
-
-      // update the statistics
-      await sqldb.queryOneRowAsync(sql.update_assessment_statisics, { assessment_id });
-    });
-  },
-
-  InstanceLogSchema,
-
-  /**
-   * Selects a log of all events associated to an assessment instance.
-   *
-   * @param {number} assessment_instance_id - The ID of the assessment instance.
-   * @param {boolean} include_files - A flag indicating if submitted files should be included in the
-   * log.
-   * @returns {Promise<Array<z.infer<typeof InstanceLogSchema>>>} the results of the log query.
-   */
-  async selectAssessmentInstanceLog(assessment_instance_id, include_files) {
-    return sqldb.queryValidatedRows(
-      sql.assessment_instance_log,
-      { assessment_instance_id, include_files },
-      InstanceLogSchema,
+
+    if (_(content.grading).has('feedback') && !_(content.grading.feedback).isObject()) {
+      throw error.makeWithData('invalid grading.feedback', { content: content });
+    }
```

Comment:
I'm considering having some of this validation be moved to a Zod schema instead, but I believe that should be done in a separate PR that focuses on that alone and does proper testing, and that is done once this file (and externalGrader.js, and possibly externalGraderCommon.js) are fully converted to TS.